src/material-experimental/mdc-slide-toggle/slide-toggle.scss

```diff
@@ -1,23 +1,30 @@
 @use 'sass:map';
-@use '@material/switch/deprecated' as mdc-switch with ($deprecated-suffix: '');
+@use '@material/switch/switch' as mdc-switch;
+@use '@material/switch/switch-theme' as mdc-switch-theme;
 @use '@material/form-field' as mdc-form-field;
 @use '@material/ripple' as mdc-ripple;
 @use '../mdc-helpers/mdc-helpers';
 @use '../../material/core/style/layout-common';
 @use '../../material/core/style/private';
 @use '../../cdk/a11y';
 
-@include mdc-switch.without-ripple($query: mdc-helpers.$mat-base-styles-query);
 @include mdc-form-field.core-styles($query: mdc-helpers.$mat-base-styles-query);
+@include mdc-switch.static-styles-without-ripple;
 
 .mat-mdc-slide-toggle {
   display: inline-block;
 
   // Remove the native outline since we use the ripple for focus indication.
   outline: 0;
 
+  .mdc-switch {
+    // MDC theme styles also include structural styles so we have to include the theme at least
+    // once here. The values will be overwritten by our own theme file afterwards.
+    @include mdc-switch-theme.theme-styles(mdc-switch-theme.$light-theme);
```

Comment:
I didn't think something like this should have been necessary. Which structural styles are in the theme? We can raise this to the MDC team if something seems wrong

Reply:
I think it's easier to look at the variables, but the tl;dr is that there a bunch of things like the track height and handle width. They aren't affected by the density, but they are still in the theme.

For reference: https://github.com/material-components/material-components-web/blob/master/packages/mdc-switch/_switch-theme.scss#L109

Reply:
Also after chatting with them on this, the idea is that the size isn't affected by the density setting in vanilla Material Design, but alternate/sub-specs could change it. I think we should just let these styles exist in the theme for now and omit them from the component's base styles.

Reply:
You can use `@material/theme/map-ext` `pick()` to help with this by only including the keys you want from the default `$light-theme`, or `split()` to remove the keys you don't want:

```scss
@include mdc-switch-theme.theme-styles(map-ext.pick(
  mdc-switch-theme.$light-theme,
  handle-height,
  handle-shape,
  handle-width,
  // ...
));
```
```scss
$ng-non-themable: list.nth(
  map-ext.split(
    mdc-switch-theme.$light-theme,
    selected-handle-color,
    // ...
  ),
  1
);
@include mdc-switch-theme.theme-styles($ng-non-themable);
```

Reply:
I've spent some time trying to get this approach to work, but for some reason it just doesn't generate any styles. For reference, here's what I'm doing:

```
$_non-theme-styles: list.nth(mdc-map-ext.split(
  mdc-switch-theme.$light-theme,
  map.keys(private-get-theme-base-map(false))...
), 1);

...
@include mdc-switch-theme.theme-styles($_non-theme-styles);
```

The idea being that we generate the excluded properties from the list of overridden properties from the theme. This way we don't have to maintain an explicit excludes list and the code will be a bit more robust against MDC introducing new properties that we don't know about.

I've tried logging out the generated map with `@debug` and it does look as I would expect so either I'm not using the theme mixin as intended, or this case isn't supported. Do you have any pointers @asyncLiz?

Here's what the result looks like:
![Angular_Material_-_Google_Chrome_2021-07-15_18-05-07](https://user-images.githubusercontent.com/4450522/125833196-dae745cd-e4f6-4a75-a5dc-07d845b3b21f.png)

Reply:
Ok gotcha, I looked more at your Sass and I understand better what's going on.

`theme-styles()` generates the "slots" for custom properties emitted by `theme()`, so every single token (a full `$theme` map) needs to be provided to create the `var()`s for those custom properties. I assumed you were using `theme-styles()` for everything, like your `color()` mixins.

```scss
.my-custom-switch-with-theme {
  @include switch-theme.theme((selected-handle-color: teal));
}

.my-custom-switch-with-theme-styles {
  @include switch-theme.theme-styles((selected-handle-color: teal));
}
```
```css
.my-custom-switch-with-theme {
  --mdc-switch-selected-handle-color: teal;
}

.my-custom-switch-with-theme-styles.mdc-switch--selected:enabled .mdc-switch__handle::after {
  background: teal;
  /* @alternate */
  background: var(--mdc-switch-selected-handle-color, teal);
}
```

So! I'm not sure on which is the desired use case (since I know Angular is moving to custom props for theming, but I'm not sure if that's happening right now for switch). If this is correct (users theming in Angular will be emitting custom properties), then calling `theme-styles()` with the full `$light-theme` is the correct thing to do. The only place the values of `$light-theme` will be used is in the fallback `var()` declarations.

If users theming the switch should be emitting selectors and classic CSS (second option), then you'll want to change your mixins to use `theme-styles()` instead of `theme()` and then do the exclude recommendation I made.

So **tl;dr** I may have led you down the wrong path... if so my bad and you can ignore me 😅

Reply:
Thank you for the explanation, that makes sense now. I had tried to do something similar on my own the first time around and got similar results, but I assumed that I had messed up somewhere along the way. I'll leave it as is for now.